### DISCUSSION TITLE
JPA 엔티티에 대한 equals 재정의 시 주의해야할 점은 무엇일까요?

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#222] [Spring Data]
뭘까용

**Comment by Kimprodp** [comment]
1. 식별자(ID)를 기준으로 비교하는 경우 엔티티가 영속화 되기 전에는 id가 Null이기 때문에 NullPointerException이 발생할 수 있다. 또한 비교하려는 두 엔티티가 모두 비영속 상태일 경우 둘 다 id가 null 이기 때문에 equals 메서드를 사용할 경우 동등하다고 판단된다. (Objects.equals는 비교 인자가 둘 다 null 일 경우 같다고 판단) 
- id를 기반으로 비교하나, 식별자가 할당되지 않은 상태일 경우 false를 반환하도록 할 수 있음

2. 엔티티 객체를 비교할 때 getClass()를 사용한다면, 엔티티는 프록시 객체로 래핑될 수 있으므로 문제가 발생할 수 있다.
(원래 클래스를 상속받는 프록시 객체를 생성하므로 instanceof() 메서드를 활용)

3. 프록시 객체랑 비교하는 경우 비교 대상이 id라면 프록시의 경우 실제 id 를 필드고 가지고 있지 않기 때문에 getId()와 같은 getter를 사용해야 함

3. hashCode 메서드도 함께 재정의할 때 식별자가 없는 경우 고정된 값으로 반환하거나 임시 값을 반환하도록 해야 함

**Comment by seunghye218** [comment]
비즈니스 키로 구현하고 해당 키가 @ManyToOne인 경우 equals()를 사용하려고 할 때 마다 지연 로딩이 발생하고 n + 1 선택 문제가 발생할 수 있습니다. 이를 피하려면 id가 선호됩니다.

JPA는 엔티티의 지연 로딩을 위해 프록시 객체를 사용할 수 있습니다. 프록시 객체의 클래스는 실제 엔티티 클래스와 다를 수 있으므로 `equals` 메서드에서 클래스를 비교할 때 주의해야 합니다. 일반적으로 프록시 객체를 올바르게 비교하려면 `instanceof`를 사용하여 타입을 확인하는 것이 좋습니다.
